### PR TITLE
[SPARK-11068][SQL] add callback to query execution

### PR DESCRIPTION
**Pull request by cloud-fan**
With this feature, we can track the query plan, time cost, exception during query execution for spark users.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @yhuai

**Comment by rxin** [comment]
Does this need to be a public interface?

If yes, can we add experimental label to it.

**Comment by rxin** [comment]
Actually if QueryExecution is not a public interface (it is in execution - I think it's incorrectly labeled developer api), this should probably be in execution as well as a semi private interface?

**Comment by cloud-fan** [comment]
retest this please.

**Comment by cloud-fan** [comment]
retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/QueryExecutionListener.scala [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+
+/**
+ * The interface of query execution listener that can be used to analyze execution metrics.
+ *
+ * Note that implementations should guarantee thread-safety as they will be used in a non
+ * thread-safe way.
+ */
+trait QueryExecutionListener {
+  def onSuccess(funcName: String, qe: QueryExecution, duration: Long)
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/QueryExecutionListener.scala]: This should be thread-safe.

File: sql/core/src/main/scala/org/apache/spark/sql/QueryExecutionListener.scala [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+  def onFailure(funcName: String, qe: QueryExecution, exception: Exception)
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/QueryExecutionListener.scala]: Could you add some java doc for all the public interface?

**Comment by cloud-fan** [comment]
retest this please.

**Comment by davies** [comment]
LGTM, waiting for tests.

**Comment by davies** [comment]
Merged into master, thanks!

**Pull request closed without merging**